From 716897df8632b211b476eb8b6539de82d90322c0 Mon Sep 17 00:00:00 2001
From: Song Zhao <song.zhao@amlogic.com>
Date: Fri, 30 Jul 2021 05:21:53 +0000
Subject: [PATCH] avsync-lib: refine pattern correction

PD#SWPL-56526

Problem:
For netflix the 2-2 pattern has PTS triplet 2970/2970/3060, with 59.96
output, each VSYNC will increase the wall clock by 1501. PTS will be
close to VYNSC point and then be apart from VSYNC and so on.

Solution:
Correct the wrong condition.

Verify:
In JIRA

Change-Id: I159f21da0c7739a7eaf14c90476dd8036f945bc0
Signed-off-by: Song Zhao <song.zhao@amlogic.com>
---
 src/pattern.c | 18 +++++++++++++-----
 1 file changed, 13 insertions(+), 5 deletions(-)

diff --git a/src/pattern.c b/src/pattern.c
index 8210c08..cc018f3 100644
--- a/src/pattern.c
+++ b/src/pattern.c
@@ -80,7 +80,7 @@ void correct_pattern(void* handle, struct vframe *frame, struct vframe *nextfram
         pts90K systime, pts90K vsync_interval, bool *expire)
 {
     struct pattern_detector *pd = (struct pattern_detector *)handle;
-    int pattern_range, expected_cur_peroid;
+    int pattern_range, expected_cur_peroid, remain_period;
     int expected_prev_interval;
     int npts = 0;
 
@@ -140,17 +140,25 @@ void correct_pattern(void* handle, struct vframe *frame, struct vframe *nextfram
 
     if (*expire) {
         if (cur_peroid < expected_cur_peroid) {
+            remain_period = expected_cur_peroid - cur_peroid;
             /* 2323232323..2233..2323, prev=2, curr=3,*/
             /* check if next frame will toggle after 3 vsyncs */
             /* 22222...22222 -> 222..2213(2)22...22 */
             /* check if next frame will toggle after 3 vsyncs */
             /* shall only allow one extra interval space to play around */
-            if (((int)(systime + (expected_prev_interval + 1) *
-                        vsync_interval - npts) >= 0) &&
-                ((int)(systime + (expected_prev_interval + 2) *
-                        vsync_interval - npts) < 0)) {
+            if (systime - frame->pts <= 90) {
+                *expire = false;
+                log_debug("hold frame for pattern: %d sys: %u fpts: %u",
+                        pd->detected, systime, frame->pts);
+            } else if (((int)(systime + (remain_period + 1) *
+                        vsync_interval - npts) <= 0) &&
+                ((int)(systime + (remain_period + 2) *
+                        vsync_interval - npts) > 0)) {
                 *expire = false;
                 log_debug("hold frame for pattern: %d", pd->detected);
+            } else {
+                log_debug("not hold frame for pattern: %d sys: %u fpts: %u s-f %u nfps: %u",
+                    pd->detected, systime, frame->pts, systime - frame->pts, npts);
             }
 
 #if 0 // Frame scattering is the right place to adjust the hold time.
-- 
2.29.2

